Uar 884 extend my account journey

VladNicol opened the pull request:
### JIRA link
[HEEDLS-884](https://softwiretech.atlassian.net/jira/software/c/projects/HEEDLS/boards/753?modal=detail&selectedIssue=HEEDLS-884)

### Description

- I changed the information displayed on MyAccount page and EditDetails page (I am failing a UI test and I am not sure why)
- I update DetailsLastChecked and CentreSpecificDetailsLastChecked in the database
- I changed the unique email validation to use `UserDataService.EmailIsInUseByOtherUser`
- I removed the password check when updating details (and I renamed `UserService.UpdateUserAccountDetailsForAllVerifiedUsers` to `UserService.UpdateUserAccountDetailsForAllUsers`)
- When I update the account details, I call `UserDataService.SetCentreEmail` with the `transaction` parameter null. This opens a new transaction inside the method. I added some code to close the transaction in case the parameter in null (I am not sure this is right). I was getting an exception in the next method that was getting called, which was `groupsService.SynchroniseUserChangesWithGroups` and it looked like the transaction not being committed was the reason.
- The 'Default' tag in the centre specific details looks bad, but I am not sure how to make it better (second screenshot)

### Screenshots
![image](https://user-images.githubusercontent.com/105208326/169981001-6c998d8b-dcc9-45a9-8268-fb3bdc04c2ab.png)
![image](https://user-images.githubusercontent.com/105208326/169981146-f8ad46aa-c0f9-4fe1-86c0-a3b71fbe0261.png)
![image](https://user-images.githubusercontent.com/105208326/169981369-4eb86afd-a9be-4c21-98c4-a946f742b846.png)
![image](https://user-images.githubusercontent.com/105208326/169981532-7941e069-ac5a-4bd4-a6db-d87e9bb602ba.png)
![image](https://user-images.githubusercontent.com/105208326/169981820-dfabf9d3-14c3-478f-b59f-b3d9425a4796.png)

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

SteveJacksonSoft commented:
@VladNicol Couple of visual things

1. There's a line at the bottom of the tables, below the profile picture and below the last registration prompt. Can we get rid of those? 
2. On the MyAccount page (not the edit details page), you're showing `Centre Email (optional)`. But there'll always be a value for that, it'll just default to the primary email if no secondary email is set. So it feels a bit strange to say (optional). Can we get rid of that?
3. It would be nicer to have the centre email `Default` tag in the table rather than beside it. You could try just inlining it with the centre email, or you might want to add a new column to the table just for the tag - have a play around and see what works best.
4. On the EditDetails page, there's no space between `Centre Email` and `(optional)`. Can you add one?

AlexJacksonDS commented on DigitalLearningSolutions.Web/Controllers/MyAccountController.cs:
```diff
@@ -56,6 +56,7 @@ public IActionResult Index(DlsSubApplication dlsSubApplication)
                     delegateUser
                 );
 
+            delegateUser.EmailAddress = userService.GetCentreEmail(User.GetUserId()!.Value, User.GetCentreId());
```
Also both the index and edit details will fail for basic users with no centre. I assume we're going to sort that out in a leter ticket @SteveJacksonSoft ?

AlexJacksonDS commented on DigitalLearningSolutions.Web/ViewModels/MyAccount/MyAccountViewModel.cs:
```diff
@@ -23,8 +24,9 @@ DlsSubApplication dlsSubApplication
-            AliasId = delegateUser?.AliasId;
             JobGroup = delegateUser?.JobGroupName;
+            CentreEmail = delegateUser?.EmailAddress;
+            DateRegistered = delegateUser?.DateRegistered?.ToString("dd-MM-yyyy");
```
Is this the correct date format? We ususally use DigitalLearningSolutions.Data.Helpers.DateHelper.StandardDateFormat

AlexJacksonDS commented on DigitalLearningSolutions.Data/Models/User/CentreAnswersData.cs:
```diff
@@ -28,6 +30,7 @@ public class CentreAnswersData
 
         public int CentreId { get; set; }
         public int JobGroupId { get; set; }
```
Job group is no longer centre specific so needs moving to the other object MyAccountDetailsData

livzorn replied:
can we remove job group from this model now that we've added it to MyAccountDetailsData? @VladNicol

VladNicol replied:
That's still used in some places, it didn't feel like part of the scope of this ticket to remove those. Should I do that?

livzorn replied:
Hmm maybe not, it does seem heavily used in that model. @AlexJacksonDS do you have an opinion?

AlexJacksonDS replied:
I guess it's probably fine to leave for now until remaining UAR work gets to it. I would expect when UAR is complete, this should be removed, so lets leave a TODO pointing to the cleanup ticket 887 that this is expected to be removed at some point.

AlexJacksonDS commented on DigitalLearningSolutions.Data/DataServices/UserDataService/UserCentreDetailsDataService.cs:
```diff
@@ -102,5 +110,23 @@ ELSE 0
+                    WHERE UserID = @userId AND CentreID = @centreId",
+                new { userId, centreId}
+            ).ToList();
+            if (emails.Count > 1)
+            {
+                throw new MultipleUserAccountsFoundException(
+                    $"User with id {userId} has several emails at centre with id {centreId}"
+                );
+            }
```
I don't think this exception is possible, there is a unique constraint on CentreID & Email (where not null), and one on CentreId & UserID, so each user can only have a single record per centre.

AlexJacksonDS reviewed (commented):
Several suggestions, particularly for the tidying up of the UserService logic we're using here so that we're actually updating the correct things at the correct time. Hopefully the comments make sense, but if they become too out of order because of github, I'll edit them to make more sense.

livzorn commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -182,46 +190,50 @@ List<DelegateUser> delegateUsers
 
         // TODO HEEDLS-887 Make sure this logic is correct with the new account structure
-        public void UpdateUserAccountDetailsForAllVerifiedUsers(
+        public void UpdateUserAccountDetailsForAllUsers(
```
This method name sounds to me like it's going to update all Users in the database. Maybe something like `UpdateUserDetailsForAllLinkedAccounts` would be clearer?

AlexJacksonDS replied:
I'm not sure about that name either, as we have no concept of linked accounts like we used to. I'd probably prefer something like `UpdateUserDetailsAndCentreSpecificDetails`

livzorn commented on DigitalLearningSolutions.Web/Controllers/MyAccountController.cs:
```diff
@@ -145,11 +139,30 @@ DlsSubApplication dlsSubApplication
                 ModelState.AddModelError(
                     nameof(MyAccountEditDetailsFormData.Email),
-                    "A user with this email address is already registered at this centre"
+                    "This email address is already in use"
```
Maybe we could add this error message to the CommonValidationErrorMessages helper where the other email-related error messages are?

VladNicol commented:
I think I addressed your comments, I hope I didn't misunderstand (or completely missed) any. I have also fixed that minor problem I had at the demo today that didn't allow me to set the centre email back to null.

I assigned the ticket to Olivia as second reviewer, but @AlexJacksonDS if you could also have another look that would be great!

livzorn commented on DigitalLearningSolutions.Web/Views/MyAccount/EditDetails.cshtml:
```diff
@@ -14,7 +14,7 @@
   <div class="nhsuk-grid-column-full">
     @if (errorHasOccurred) {
-      <vc:error-summary order-of-property-names="@(new[] { nameof(Model.FirstName), nameof(Model.LastName), nameof(Model.Email), nameof(Model.HasProfessionalRegistrationNumber), nameof(Model.ProfessionalRegistrationNumber), nameof(Model.Password), nameof(Model.JobGroupId), nameof(Model.Answer1), nameof(Model.Answer2), nameof(Model.Answer3), nameof(Model.Answer4), nameof(Model.Answer5), nameof(Mod ...
+      <vc:error-summary order-of-property-names="@(new[] { nameof(Model.FirstName), nameof(Model.LastName), nameof(Model.Email), nameof(Model.JobGroupId), nameof(Model.HasProfessionalRegistrationNumber), nameof(Model.ProfessionalRegistrationNumber), nameof(Model.CentreEmail), nameof(Model.Answer1), nameof(Model.Answer2), nameof(Model.Answer3), nameof(Model.Answer4), nameof(Model.Answer5), nameof( ...
```
Is it possible to break this over multiple lines?

VladNicol replied:
Yeah, the auto formatter put that on one line

SteveJacksonSoft replied:
A nice way around would be to pull the `new[] {...}` out into a variable. Then the autoformatter would break it into separate lines I think.

livzorn commented on DigitalLearningSolutions.Web/Views/MyAccount/EditDetails.cshtml:
```diff
@@ -53,35 +53,74 @@
-          default-option="Select a job group"
-          select-list-options="@Model.JobGroups" />
+
+        @{ var centreEmailHintText = "If you supply a centre email, all notifications relating to your activity at this centre will be sent to this email address. Your primary email address will still be used for logging in and notifications for managing your account."; }
```
Can we break this hint text over multiple lines?

livzorn reviewed (commented):
Nice work, a few minor things left

AlexJacksonDS commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -182,46 +190,50 @@ List<DelegateUser> delegateUsers
+                    delegateUsers.SingleOrDefault(u => u.Id == myAccountDetailsData.DelegateId);
 
-                if (oldDelegateDetails != null && centreAnswersData != null)
+                if (oldDelegateDetails != null)
```
We're still doing some unnecessary stuff here. Why do we need to fetch the oldDelegateUserDetails from the database when we've already fetched them from the database by user ID earlier? I think we should try to get the delegate user out of the previous list by id, and then only update the delegate account if it is found.

AlexJacksonDS replied:
@VladNicol In case you miss this comment that I've unresolved

VladNicol replied:
I kept that `oldDelegateUserDetails` because `userDataService.GetDelegateAccountsByUserId` returns a `IEnumerable<DelegateAccount>`, while `groupsService.SynchroniseUserChangesWithGroups` takes as argument a `DelegateUser`. Do we have a nice way to map a `DelegateAccount` to a `DelegateUser`? They have almost the same fields, in cases like this I am used to AutoMapper but we don't seem to use that in the project.

AlexJacksonDS replied:
We don't have a way to map them, but I do wonder whether changing the group synchronization so that it works with the Delgate accounts would be something in scope here. Otherwise I do think it would be good to at some point at least, since I expect we will probably end up with DelegateAccounts in the other place this is called. @SteveJacksonSoft Do we have a ticket to resolve this bit of functionality specifically?

VladNicol replied:
I don't think there is a nice and easy way around this, at least until `JobGroupId` is removed from `CentreAnswersData`. We have `GroupsService.SynchroniseUserChangesWithGroups` that calls `LinkedFieldHelper.GetLinkedFieldChanges(CentreAnswersData oldAnswers, CentreAnswersData newAnswers, ...)`.

I think we have to first remove `JobGroupId` from `CentreAnswersData` and then create a `DelegateAccount.GetCentreAnswersData` similar to `DelegateUser.GetCentreAnswersData`

SteveJacksonSoft replied:
The group service method still needs the same data, it's just that that data isn't all in the same place anymore, so some renaming / reshuffling needs to be done in the models. I'll add comments in appropriate places.

AlexJacksonDS commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -193,77 +201,74 @@ List<DelegateUser> delegateUsers
-                );
+            var delegateUsers = userDataService.GetDelegateAccountsByUserId(userId).ToList();
+
+            var detailsLastChecked = DateTime.Now;
```
Didn't notice this before, but we should be using ClockService.UtcNow for this

AlexJacksonDS commented on DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/EditDelegateController.cs:
```diff
@@ -75,11 +75,11 @@ public IActionResult Index(EditDelegateFormData formData, int delegateId)
                 ModelState.AddModelError(
                     nameof(EditDetailsFormData.Email),
-                    "A user with this email address is already registered at this centre"
+                    "This email address is already in use"
```
This also needs to be the common error message

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -193,77 +201,75 @@ List<DelegateUser> delegateUsers
-        public bool NewEmailAddressIsValid(string emailAddress, int? adminUserId, int? delegateUserId, int centreId)
+        public bool NewEmailAddressIsValid(
+            string emailAddress,
+            int? adminUserId,
+            int? delegateUserId,
+            int? userId = null
```
I'm a bit scared that this method is allowing people to pass in adminId and delegateId and userId.

What happens if someone decides to call this method with a userId and a delegateId which correspond to different users?

We could avoid this by disallowing any calling code from passing in adminUserId or delegateUserId, demanding that it gives us a userId and centreId instead (which makes the implementation of this method simpler).
Can we do this instead?

VladNicol replied:
I don't think I need the centre id, I can just return `!userDataService.EmailIsInUseByOtherUser(userId, emailAddress)`, or am I missing something?

AlexJacksonDS replied:
For checking UserCentreDetails.

SteveJacksonSoft replied:
Use `userDataService.EmailIsInUseByOtherUser(userId, emailAddress)` instead of this method.

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -193,77 +201,75 @@ List<DelegateUser> delegateUsers
+            int userId,
             MyAccountDetailsData myAccountDetailsData,
-            CentreAnswersData? centreAnswersData = null
+            CentreAnswersData centreAnswersData
```
Split this into two params
* `string? centreEmail`
* `DelegateDetailsData? delegateDetailsData`

Such that if delegateDetailsData is non-null then the user is assumed to have a delegate account. 
Just noticed delegateId is on MyAccountDetailsData - might make sense to port that across to delegateDetailsData.

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -193,77 +201,75 @@ List<DelegateUser> delegateUsers
+                groupsService.SynchroniseUserChangesWithGroups(
+                    oldDelegateUserDetails!,
+                    myAccountDetailsData,
+                    centreAnswersData
```
Here call constructor on ~CentreAnswersData~ RegistrationFieldAnswers to combine `delegateDetailsData` and `jobGroup`

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Models/User/CentreAnswersData.cs:
```diff
@@ -8,6 +8,7 @@ public class CentreAnswersData
         public CentreAnswersData(
```
Rename this to `RegistrationFieldAnswers` and add a class `DelegateDetailsData`, and a constructor to turn `DelegateDetailsData` into `RegistrationFieldAnswers`.

Hopefully you can remove the existing constructor too.
I suggested that one should inherit the other, but that actually feels wrong, conceptually. Adding to the `DelegateDetailsData` model won't necessarily add to the registration fields we can use to generate groups.

(Can't comment on the class declaration 😢)

SteveJacksonSoft commented on DigitalLearningSolutions.Data/DataServices/UserDataService/UserCentreDetailsDataService.cs:
```diff
@@ -73,6 +77,11 @@ ELSE 0
+
+            if (transactionShouldBeClosed)
+            {
+                transaction.Commit();
```
Good catch!

VladNicol commented on DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/EditDelegateController.cs:
```diff
@@ -75,11 +75,11 @@ public IActionResult Index(EditDelegateFormData formData, int delegateId)
             }
 
-            if (!userService.NewEmailAddressIsValid(formData.Email!, null, delegateId, centreId))
+            if (!userService.NewEmailAddressIsValid(formData.Email!, null, delegateId, User.GetUserId()!.Value))
```
The userId is wrong here, it should be the id of the user whose details are being edited

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/GroupsService.cs:
```diff
@@ -38,8 +38,8 @@ public interface IGroupsService
             DelegateUser delegateAccountWithOldDetails,
-            AccountDetailsData newDelegateDetails,
-            CentreAnswersData newCentreAnswers
+            AccountDetailsData a,
```
probably a mistake

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -35,17 +31,13 @@ List<DelegateUser> delegateUsers
-            CentreAnswersData centreAnswersData
+            DelegateDetailsData? delegateDetailsData,
+            string? centreEmail,
+            int? centreId
```
I think it makes sense to put centreId in DelegateDetailsData - it is in the DelegateAccounts table after all.

AlexJacksonDS replied:
Users can have a centre email without a delegate account, we need it out

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -218,58 +186,47 @@ CentreAnswersData centreAnswersData
-                centreAnswersData.CentreId,
-                centreAnswersData.CentreEmail
-            );
+            if (centreId != null)
```
Surely this should be if centreEmail not null?

AlexJacksonDS replied:
No, we need to be able to clear centre email

SteveJacksonSoft replied:
Aha, gotcha. Guessing you've left the centreId nullable in preparation for the centre-independent EditDetails journey.
I think we want to be using a new method for that anyway, so I think it'd be better to have centreId not-nullable.

SteveJacksonSoft replied:
For now it's fine if we don't save anything when User.GetCentreId() returns null.

SteveJacksonSoft commented on DigitalLearningSolutions.Data/Services/UserService.cs:
```diff
@@ -218,58 +186,47 @@ CentreAnswersData centreAnswersData
+            if (delegateDetailsData != null)
             {
-                var delegateId = myAccountDetailsData.DelegateId!.Value;
                 // TODO HEEDLS-887 Remove this call once groupsService.SynchroniseUserChangesWithGroups is refactored
```
If we're ever going to refactor that method, now is probably the time to do it.

SteveJacksonSoft replied:
Just working out if we actually have sufficient motivation to.

SteveJacksonSoft replied:
I don't think we do actually. Even if we do refactor it to use a `Delegate` rather than a `DelegateUser` you'll still have to call the db to get the delegate's old registration field responses. And this line is all it would change (along with the line in the other place it's called I guess).

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/EditDelegateController.cs:
```diff
@@ -75,7 +75,8 @@ public IActionResult Index(EditDelegateFormData formData, int delegateId)
             }
 
-            if (!userService.NewEmailAddressIsValid(formData.Email!, null, delegateId, User.GetUserId()!.Value))
+            // TODO HEEDLS-887 Fix this so that it is bassing the correct User ID, that of the delegate not the logged in admin
```
@GeorgeBlackwell Is this being done in your ticket? We can talk about it tomorrow.

AlexJacksonDS replied:
This is really https://softwiretech.atlassian.net/browse/HEEDLS-951 I don't see why @GeorgeBlackwell would be changing this on his ticket? I'll mark it with the correct todo number
